server/src/main/scala/org/http4s/server/middleware/Jsonp.scala

```diff
@@ -26,24 +26,23 @@ object Jsonp {
 
-  def apply[F[_]: Monad](callbackParam: String)(@deprecatedName('service) routes: HttpRoutes[F])(
-      implicit W: EntityEncoder[F, String]): HttpRoutes[F] =
+  def apply[F[_]: Monad, G[_]: Monad](callbackParam: String)(@deprecatedName('service) http: Http[F, G]): Http[F, G] =
     Kleisli { req =>
       req.params.get(callbackParam) match {
         case Some(ValidCallback(callback)) =>
-          routes(req).map { response =>
+          http.map { response =>
             response.contentType.map(_.mediaType) match {
               case Some(MediaType.application.json) =>
                 jsonp(response, callback)
-              case _ =>
-                response
+              case _ => response
             }
-          }
+          }.apply(req)
+
         case Some(invalidCallback) =>
           logger.warn(s"Jsonp requested with invalid callback function name $invalidCallback")
-          OptionT.pure(Response[F](Status.BadRequest).withEntity(s"Not a valid callback name."))
-        case None =>
-          routes(req)
+          http.map(response => response.withStatus(Status.BadRequest).withEntity(s"Not a valid callback name.")).apply(req)
```

Comment:
This applies the `http` only to throw it away.  I think we want to add a `G ~> F` parameter, like `ChunkAggregator` has.  And then we can use that to lift the bad request response into `F` without calling the `http`.